web/app/js/components/ScatterPlot.jsx

```diff
@@ -128,49 +162,58 @@ export default class ScatterPlot extends React.Component {
       .text("Success rate");
   }
 
+  getNearbyDatapoints(x, data) {
+    // return nodes that have nearby x-coordinates
+    let xMargin = 30; // todo move constant
+    let x0 = this.xScale.invert(x - xMargin);
+    let x1 = this.xScale.invert(x + 2 * xMargin);
+
+    return  _.filter(data, d => {
+      let latency = getP99(d);
+      return latency <= x1 && latency >= x0;
+    });
+  }
+
   updateGraph() {
-    let plotData = _.reduce(this.props.data, (mem, datum) => {
-      if(!_.isNil(datum.scatterplot.success) && !_.isNil(datum.scatterplot.latency)) {
-        mem.push(datum.scatterplot);
-      }
-      return mem;
-    }, []);
-    this.updateScales(plotData);
+    let plotData = this.props.data;
+
     this.scatterPlot = this.svg.selectAll(".dot")
-      .data(plotData)
+      .data(plotData);
 
-    this.labels = this.svg.selectAll(".dot-label")
-      .data(plotData)
+    this.scatterPlot.exit().remove();
 
     this.scatterPlot
       .enter()
-        .append("circle")
-        .attr("class", "dot")
-        .attr("r", circleRadius)
+      .append("circle")
+      .attr("class", "dot")
+      .attr("r", circleRadius)
       .merge(this.scatterPlot) // newfangled d3 'update' selection
-        .attr("cx", d => this.xScale(_.get(d, ["latency", "P99", 0, "value"])))
-        .attr("cy", d => this.yScale(d.successRate))
-        .style("fill", d => successRateColorScale(d.successRate))
-        .style("stroke", d => successRateStrokeColorScale(d.successRate))
-        .on("mousemove", d => {
-          let sr = metricToFormatter["SUCCESS_RATE"](d.successRate);
-          let latency = metricToFormatter["LATENCY"](_.get(d, ["latency", "P99", 0, "value"]));
-          this.tooltip
-            .style("left", d3.event.pageX - 50 + "px")
-            .style("top", d3.event.pageY - 70 + "px")
-            .style("display", "inline-block") // show tooltip
-            .text(`${d.name}: (${latency}, ${sr})`);
-        })
-        .on("mouseout", () => this.tooltip.style("display", "none"));
-
-    this.labels
-      .enter()
-        .append("text")
-        .attr("class", "dot-label")
-      .merge(this.labels)
-        .text(d => d.name)
-        .attr("x", d => this.xScale(_.get(d, ["latency", "P99", 0, "value"])) - circleRadius)
-        .attr("y", d => this.yScale(d.successRate) - 2 * circleRadius)
+      .attr("cx", d => this.xScale(getP99(d)))
+      .attr("cy", d => this.yScale(d.successRate))
+      .style("fill", d => successRateColorScale(d.successRate))
+      .style("stroke", d => successRateColorScale(d.successRate));
+
+    this.overlay
+      .on("mousemove", () => {
+        let currXPos = d3.mouse(d3.select("rect").node())[0];
+        this.verticalHighlight.attr("transform", "translate(" + currXPos + ", 0)");
```

Comment:
i noticed when i move the mouse left, the vertical highlight smoothly follows, but moving the mouse right, the highlight stays put until i hit its right edge, and then it jumps a full width?

Reply:
whooooaaaa that's interesting. lemme look into it.